pkg/dedup/dedup_linux.go

```diff
@@ -0,0 +1,139 @@
+package dedup
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"sync"
+	"syscall"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+type dedupFiles struct {
+	lock          sync.Mutex
+	dev           uint64
+	visitedInodes map[uint64]struct{}
+}
+
+func newDedupFiles() (*dedupFiles, error) {
+	return &dedupFiles{
+		visitedInodes: make(map[uint64]struct{}),
+	}, nil
+}
+
+func (d *dedupFiles) recordInode(dev, ino uint64) (bool, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	// record the device used by the first inode, do not allow
+	// recording inodes from different devices
+	if d.dev == 0 {
+		d.dev = dev
+	} else if d.dev != dev {
+		logrus.Warningf("Ignored inode from different device: %d (expected %v)", dev, d.dev)
```

Comment:
I’m tempted to read the `%d` vs. `%v` difference as an indication that something non-obvious is going on; if it isn’t the case, please use the same directive for both.